### PR TITLE
feat: Allow undo/redo in form dialog schema builder

### LINKED ISSUE #4327: Allow undo/redo in form dialog schema builder
Enable users to undo/redo their actions while building a form dialog schema

### PR DESCRIPTION
**Pull request by hatpick**
## Description

Enable users to undo/redo their actions while building a form dialog schema

## Task Item

closes #4327 

## Screenshots

N/A

### CODE REVIEW

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx [hunk 1]
```diff
@@ -0,0 +1,173 @@
+// Copyright (c) Microsoft Corporation.
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:1]: Nit: UndoRoot reads like a command.  RecoilRootWithUndo?  UndoableRecoilRoot? There may not be a better name...

File: Composer/packages/form-dialogs/src/components/FormDialogPropertiesEditor.tsx [hunk 2]
```diff
@@ -78,6 +80,7 @@ export const FormDialogPropertiesEditor = React.memo((props: Props) => {
 
   const schemaIdRef = React.useRef<string>(schema.id);
 
+  const { undo, redo, canUndo, canRedo } = useUndo();
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/components/FormDialogPropertiesEditor.tsx:83]: You could return undo as undefined when you can't undo but maybe having the booleans is semantically clearer.

File: Composer/packages/form-dialogs/src/components/common/CommandBarUpload.tsx [hunk 3]
```diff
@@ -27,6 +27,7 @@ export const CommandBarUploadButton = (props: Props) => {
   const onChange = () => {
     if (inputFileRef.current.files) {
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/components/common/CommandBarUpload.tsx:28]: Should this be current?.files  - current for a ref can be null right (although for a UI callback very unlikely)

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+// Licensed under the MIT License.
+
+/* eslint-disable @typescript-eslint/no-explicit-any */
+
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:5]: This is a significant amount of work to support undo.  Maybe we can lift this up to shared across Composer for other plug-ins and extensions.

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+  }
+  return atomMap;
+};
+
+const didAtomMapsChange = (prevMap: AtomMap, currMap: AtomMap) => {
+  if (prevMap.size !== currMap.size) {
+    return true;
+  }
+
+  for (const key of prevMap.keys()) {
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:31]: Is it possible for the sizes to be the same, but then currMap has a key that prevMap does not?  You could use lodash intersection/union methods to determine if there are any key differences.

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+  return current.map((pendingSnap) => {
+    for (const [atom, loadable] of atomMap.entries()) {
+      if (loadable.state === 'hasValue') {
+        pendingSnap.set(atom, loadable.contents);
+      }
+    }
+  });
+};
+
+type ContextState = {
```
Comment (GeoffCoxMSFT) [inline Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:73]: Do you want to export the ContextState type so that callers can useContext with it?

**Review by GeoffCoxMSFT — DISMISSED** [review]

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx (hunk 4 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:5]: It's pretty much this one file.

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx (hunk 1 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:1]: I think it's good enough.

File: Composer/packages/form-dialogs/src/components/common/CommandBarUpload.tsx (hunk 3 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/components/common/CommandBarUpload.tsx:28]: If the ref is not set that's a major bug in react.

File: Composer/packages/form-dialogs/src/undo/UndoRoot.tsx (hunk 5 above)
Comment (hatpick) [reply Composer/packages/form-dialogs/src/undo/UndoRoot.tsx:31]: Good catch, I added that there when the trackedAtom size was fixed.

**Review by GeoffCoxMSFT — APPROVED** [review]

**Pull request merged**